community/detectors/apache_http_server_cve_2021_41773/src/test/java/com/google/tsunami/plugins/detectors/nacos/auth_bypass_lte140/ApacheHttpServerVulnDetectorTest.java

```diff
@@ -0,0 +1,128 @@
+import okhttp3.mockwebserver.RecordedRequest;
+import org.junit.After;
+import org.junit.Before;
+import org.junit.Test;
+
+public class ApacheHttpServerVulnDetectorTest {
+
+  private final FakeUtcClock fakeUtcClock =
+      FakeUtcClock.create().setNow(Instant.parse("2020-01-01T00:00:00.00Z"));
+
+  @Inject
+  private ApacheHttpServerVulnDetector detector;
+
+  private MockWebServer mockWebServer;
+
+  @Before
+  public void setUp() {
+    mockWebServer = new MockWebServer();
+  }
+
+  @After
+  public void tearDown() throws IOException {
+    mockWebServer.shutdown();
+  }
+
+  private void createInjector() {
+    Guice.createInjector(
+            new FakeUtcClockModule(fakeUtcClock),
+            new HttpClientModule.Builder().build())
+        .injectMembers(this);
+  }
+
+  @Test
+  public void testPathTraversalAndDisclosure()
+      throws IOException {
+    createInjector();
+    mockWebServer.setDispatcher(new VulnerableEndpointDispatcher());
+    mockWebServer.start();
+
+    ImmutableList<NetworkService> httpServices = ImmutableList.of(
+        NetworkService.newBuilder()
+            .setNetworkEndpoint(
+                forHostnameAndPort(mockWebServer.getHostName(), mockWebServer.getPort()))
+            .setTransportProtocol(TransportProtocol.TCP)
+            .setServiceName("http")
+            .build());
+
+    TargetInfo targetInfo = buildTargetInfo(forHostname(mockWebServer.getHostName()));
+    DetectionReportList detectionReports = detector.detect(targetInfo, httpServices);
+
+    assertThat(detectionReports.getDetectionReportsList()).containsExactly(
+        detector.buildDetectionReport(targetInfo, httpServices.get(0)));
+    assertThat(mockWebServer.getRequestCount()).isEqualTo(1);
+  }
+
+  @Test
+  public void testNoPathTraversalAndDisclosure()
+      throws IOException {
+    createInjector();
+    mockWebServer.setDispatcher(new SafeEndpointDispatcher());
+    mockWebServer.start();
+
+    ImmutableList<NetworkService> httpServices = ImmutableList.of(
+        NetworkService.newBuilder()
+            .setNetworkEndpoint(
+                forHostnameAndPort(mockWebServer.getHostName(), mockWebServer.getPort()))
+            .setTransportProtocol(TransportProtocol.TCP)
+            .setServiceName("http")
+            .build());
+
+    TargetInfo targetInfo = buildTargetInfo(forHostname(mockWebServer.getHostName()));
+    DetectionReportList detectionReports = detector.detect(targetInfo, httpServices);
+
+    assertThat(detectionReports.getDetectionReportsList()).isEmpty();
+    assertThat(mockWebServer.getRequestCount()).isEqualTo(1);
+  }
+
+
+  static final class VulnerableEndpointDispatcher extends Dispatcher {
+
+    @Override
+    public MockResponse dispatch(RecordedRequest recordedRequest) {
+      return new MockResponse().setResponseCode(HttpStatus.OK.code())
+          .addHeader("Server", "Apache/2.4.49 (Unix)");
+    }
+  }
+
+  static final class SafeEndpointDispatcher extends Dispatcher {
```

Comment:
These 2 `Dispatcher` class can be `private`.

Reply:
done.